torch_geometric/nn/conv/heat_pyg.py

```diff
@@ -0,0 +1,187 @@
+        self.node_update_emb = nn.Linear(self.edge_attr_emb_size + self.node_emb_size, self.out_channels, bias=False) 
+
+        ## Attention
+        self.attention_nn = nn.Linear(self.edge_attr_emb_size + self.edge_type_emb_size + 2*self.node_emb_size, 1*self.heads, bias=False)
+        self.leaky_relu = nn.LeakyReLU(0.2)
+        self.soft_max = nn.Softmax(dim=1)
+    
+    def set_node_emb(self):
+        ''' assume that different nodes have the same dimension, but different vector space. '''
+        for node_type in self.node_type_set:
+            exec('self.{}_node_feat_emb = nn.Linear(self.in_channels_node, self.node_emb_size, bias=False)'.format(node_type))
+    
+    def set_edge_emb(self):
+        ''' assume that different edges have different types but contains the same kind of attributes. '''
+        # for edge_type in self.edge_type_set:
+        #     exec('self.{}_edge_type_emb = nn.Linear(self.in_channels_edge_type, self.edge_type_emb_size, bias=False) '.format(edge_type))
+        self.edge_type_emb = nn.Linear(self.in_channels_edge_type, self.edge_type_emb_size, bias=False) 
+        self.edge_attr_emb = nn.Linear(self.in_channels_edge_attr, self.edge_attr_emb_size, bias=False) 
+
+    def embed_nodes(self, node_features, node_type_list):
+        ''' embed node attributes according to their types. '''
+        emb_node_features = torch.zeros(node_features.shape[0], self.node_emb_size).to(node_features.device)
+        for node_type in self.node_type_set:
+            node_mask = retrieve_masks(node_type_list, wanted_types=[node_type])
+            exec('emb_node_features[node_mask] = self.{}_node_feat_emb(node_features[node_mask])'.format(node_type))
+        return emb_node_features
+    
+    def embed_edges(self, edge_attrs, edge_types):
+        ''' embed edge attributes and edge types respectively. '''
+        emb_edge_attributes = self.leaky_relu(self.edge_attr_emb(edge_attrs))
+        emb_edge_types = self.leaky_relu(self.edge_type_emb(edge_types))
+        return emb_edge_attributes, emb_edge_types
+    
+    def forward(self, node_feats, edge_index, edge_attr, edge_type, node_type_list, edge_type_list):
+        """
+        Args:
+            node_f:[num_node, in_channels_node]
+            edge_index: [2, number_edge]
+            edge_attr: [number_edge, len_edge_attr]
+            edge_type: [number_edge, len_edge_type]
+            node_type_list: [ type of each node ]
+            edge_type_list: [ type of each edge ]
+        """
+        # Node-type specific transformation
+        node_emb = self.embed_nodes(node_feats, node_type_list)
+
+        # Edge attribute and type transformation
+        edge_attr_emb, edge_type_emb = self.embed_edges(edge_attr, edge_type)
+
+        # Edge-enhanced masked attention
+        emb_edge_feat = torch.cat((edge_attr_emb, edge_type_emb), dim=1) # e_{i,j} in heat paper
+
+        src_node_emb = node_emb[edge_index[0]]
```

Comment:
We should make use of the `MessagePassing` interface to implement this, i.e. in `self.propagate`.